entities.py

```diff
@@ -251,28 +267,40 @@ def _compute_max_hp(self):
         # 25% STR, 75% CON, 100 BASE
         hp_str = self.strength * self.level * 0.25
         hp_con = self.constitution * self.level * 0.25
-        self.max_hp = hp_str + hp_con + 100
+        return hp_str + hp_con + 100
 
     def _compute_max_mp(self):
         # 100% INT, 50 BASE
-        self.max_mp = self.intelligence * self.level + 50
+        return self.intelligence * self.level + 50
 
     def _compute_ranged_damage(self):
         # 100% DEX
-        self.ranged_dmg = self.dexterity * self.level * 0.5
+        return self.dexterity * self.level * 0.5
 
     def _compute_magical_damage(self):
         # 100% INT
-        self.magical_dmg = self.intelligence * self.level * 0.33
+        return self.intelligence * self.level * 0.33
 
     def _compute_physical_damage(self):
         # 100% STR
-        self.physical_dmg = self.strength * self.level * 0.66
+        return self.strength * self.level * 0.66
 
     def _compute_dodge_rate(self):
         # 100% DEX, assumed max level = 50
-        self.dodge_rate = (self.dexterity * self.level) / 500
+        return (self.dexterity * self.level) / 500
 
     def _compute_crit_rate(self):
         # 100% LCK, assumed max level = 50
-        self.crit_rate = (self.luck * self.level) / 500
+        return (self.luck * self.level) / 500
+
+    def _recompute_stats(self):
```

Comment:
that's a lot of useless computing, but might as well make it into a decorator and decorate setters that would trigger recomputing with it, no?

e.g.
```python
@recompute_stats
@constitution.setter
def constitution(self, val):
    ...
```

Reply:
Yeah I know it adds useless computing, but I actually did it on purpose. It really doesn't matter since parameters are computed with just a couple of arithmetic operations, and they would only be recomputed when a stat changes so it doesn't add a lot of overhead. Also, this way the setters don't have to know about parameter dependency. If this is ever a performance bottleneck (which it won't) then sure I'll change it back, but until then I actually think this keeps things simpler.

The decorator is a good idea tho, will work on that.

Reply:
no actually I changed my mind, it's a nice idea since the decorator makes recomputing far easier and we know which stats are computed and which ones trigger the recomputing and as you said, it's unlikely that this will ever be a performance bottleneck